### ISSUE TITLE
Multiple ajaxed-in Lightbox and/or Overlay instances on a page = Extreme lag or page kill

### ISSUE THREAD

**Issue by markbelanger** [wet-boew/wet-boew#6931]
Is there a way to avoid or to store the call to the i18n and Magnific pop-up JS file when using the Lightbox and/or Overlay plug-ins multiple times on a page? 

I'm using these plug-ins to ajax-in dynamic content on search results records, that contain multiple different Source snippets within each individual record. Search results are limited to 100 records on a page, but each records can contain up to 10 sources (which can equate to a very large amount of Sources on one full page). 

When using the Overlay plug-in to ajax-in the source content, the page lags so much that it often times kills the browser altogether. I looked at what FIrebug is trying to process with the JS, and I can see that it's constantly hitting the i18n language file and Magnific Pop-up JS file to process every single time the required functions to make each overlay render what it needs. 

Is there any way to store the required i18n and Magnific script and not have it process its function and sub-routines on every single call? Something that can declare the same parameters to all overlay instances, seeing as that page does not change language per instance of overlay, not does the Magnific pop-up script requirement change for each instance. All the overlays contain the same parameters, just different content.

I'd really like to take advantage of the Overlay/Lightbox plug-in for a better user experience, but killing the browser or slowing down the load is obviously not an option. 

I want to avoid having a link to a separate page for every Source instance, as the content of each source is minimal, and not worth the extra click or page change. I want the user to quickly have access to their search results and avoid back-and-forth clicking between results and sources.

**Comment by pjackson28** [comment]
Do you have a page we can look at? We do use caching and store the calls wherever possible and haven't had any performance issues that we can tell. For instance the i18n file is loaded once then cached as an object on the Windows object so it can be used by all instances and the Magnific Popup JS file is loaded only once through a JS loader. Do you having caching enabled or disabled?

For instance, here is where we initialize the i18n and Magnific Popup callback object only once for the Lightbox plugin: https://github.com/wet-boew/wet-boew/blob/master/src/plugins/lightbox/lightbox.js#L119-L136. We welcome any other suggestions for improving performance.

By the way, have you timed to see where all the time is spent processing? If so, please share the results. From what you describe, what you are AJAXing in could be a big factor in your performance issues. How you AJAX in that information (number of calls and where you put the returned data) can have a huge impact. For instance if you have dozens of concurrent calls and then manually insert each record into the DOM one at a time then you will drive the page into the ground (each AJAX call and each DOM hit is very expensive performance wise). If you are not doing so already, you should minimize the number of AJAX calls (pull the data in and then filter/sort) and only insert the data into the page once (single DOM hit) rather than repeatedly.

**Comment by LaurentGoderre** [comment]
If your server is not correctly configured to send the client-side caching headers, the dependency loader will not work correctly and will send multiple useless requests.

As you can see here from the GitHub hosted examples, the resources are cached properly 
![caching](https://cloud.githubusercontent.com/assets/1090826/8621809/a020d8fe-26f5-11e5-8acb-6bdd95dbec74.PNG)

That being said, the i18n is only getting called once in the script so there might something funky with the code you are using. Is it possible that you are not retrieving fragments but whole pages? You might be ajaxing in a new reference to the WET javascript which would explain the multiple calls

**Comment by markbelanger** [comment]
Thank you for your quick replies gentlemen. Much appreciated.

Here are a few answers to your questions:

1. The application in it's WET4 format currently resides on one of our DEV servers, which is not public facing, and has a specific IP access range. I will be providing as much code or screenshots as required to try and help identify the issue.

2. Yes, I have looked at Firebug timestamps for all the files and event loads. I can can clearly see where the delays are happening. I will post Firebug screenshots which display the state and load times of all files, including the page before and after results.

3. The i18n is only called once, but takes several seconds to load. It only completely loads once all records are returned on the page, which is what lead me to believe its (or the Overlay plug-in) sub routines need to be initiated multiple times in relation to all the elements that use it.

4. I will investigate the server caching possibly being the issue. However, I get the 304 for all the files on the domain, just like you illustrated in your screenshot. I will support this with upcoming screenshots.

5. The ajaxed-in Source files are only fetched when called upon. They do not contain the template, WET, JS or any other page elements. The only contain a panel with a header and a paragraph or two in the panel body.

As mentioned, screenshots to come as soon as possible. In the meantime, thanks again for your help and guidance. Much appreciated.

**Comment by markbelanger** [comment]
As promised, here are some Firebug screenshots, which display different states of the search. You can clearly see where the JS files are hammering the page and slowing it down.

1. Before the search, no results on page.
![before-search](https://cloud.githubusercontent.com/assets/3777175/8649550/5dfe57da-2939-11e5-9171-8932cca66df7.png)

2. After search. Showing 2 results.
![after-search-2results](https://cloud.githubusercontent.com/assets/3777175/8649562/89e69dbc-2939-11e5-9314-fe008461b5a7.png)

3. Perform a different search. Search returns 5 results. Page starts to lag a bit.
![after-search-5results](https://cloud.githubusercontent.com/assets/3777175/8649573/a37af41c-2939-11e5-9610-dbd411defcfd.png)

4. Perform another different search. Search returns 30 results. 
![after-search-30results](https://cloud.githubusercontent.com/assets/3777175/8649582/b2464d66-2939-11e5-9da3-3b499422cbf1.png)

At this point, browser is getting chocked pretty hard. Script debugger prompt appears. Scripts are unable to continue. Page is unusable at this point.
![unresponsive-script-prompt](https://cloud.githubusercontent.com/assets/3777175/8649607/e33b7b80-2939-11e5-8d8f-fb7ce525ab0c.png)

**Comment by pjackson28** [comment]
Looks like caching is definitely off as you are getting 200s for several files multiple times (jquery.min.js, en.min.js, jquery.magnific). The loader we use is dependent upon caching and if you have it off you will end up with multiple uncached requests for the same file. I also noticed that you have some very heavy frameworks that do a lot of expensive DOM manipulations (e.g., jQuery UI) so that will really burden the page (we stopped using jQuery UI quite a while back because it was so hard on performance, and because it does not play well with other libraries). I also noticed that you spend a lot of time in gatrack.min.js which can get really performance heavy as well if you are not careful (since it can make a lot of DOM changes),

So what I would recommend is the following:
1. Turn caching on. You should not see the same file multiple times with a 200 code.
2. Turn off all non-WET JS files.
3. Check the performance again.

If you are still having performance problems then you should then you should look into how you are doing your searches. If you are doing 30 separate requests then that could quickly bottleneck so you should combine the requests together as much as possible to keep the number of requests low. Try pulling back a larger data set and filtering locally rather than multiple individual calls. Try to find a good balance between download size and number of calls because two much on either side can have a big impact on performance. Also look into how the search results are being used. You should build up all the HTML changes into one big string and then write to the DOM once rather than multiple individual writes to the DOM (since DOM changes are very expensive).

**Comment by markbelanger** [comment]
Sorry for the delay on follow-up. Tried all suggestions, still have a significant lag in performance. Investigating back-end and server-side processes. Will follow-up once thorough investigation is complete. Thanks for your guidance. Much appreciated.

**Comment by pjackson28** [comment]
I would look for much more lightweight non-jQuery UI approaches as relying upon jQuery UI for what you are trying to achieve is likely a big factor in why you are getting performance issues (since jQuery UI can be quite expensive performance-wise). For switchclass in particular, you should be using CSS transitions directly rather than relying upon jQuery UI switchclass for this. We use CSS transitions for animations in WET rather than any of the animations in jQuery because of the large performance costs of JS-based animation (especially on mobile devices where JS-based animation can grind things to a halt).

**Comment by markbelanger** [comment]
Thanks Paul. I'll look into an alternative to switchclass as per your advice. I had broken down jquery UI to just the few functions I was using, and built a new minimal library with what I had left, but it was indeed still killing my pages that return many results. Thanks for the advice. To be continued.

**Comment by markbelanger** [comment]
Update:
1. Successfully removed any reference to jquery-ui and rewrote switchclass to use addClass/removeClass available in jquery WET template.
2. Grouped and minified all custom functions an styles into one single file (one JS and one CSS). These include autocomplete (to call upon a large data source, not a basic list of text options like the WET example), chosen (jquery DDL) and loader functions not available "out-of-the-box" via WET4 or Bootstrap.
3. Successfully reduced page load time by several seconds compared to initial screenshots above. Alos no more complete page/browser crash.
4. Removed reference calling gattrack to see if it speeds page load time -- it does. Left commented out for the time being until process of elimination is complete.
5. Still experiencing unwanted page lag. Not as bad as before, but far from optimal (not release worthy for international user community). Defnitely due to javascript -- when JS disabled, page load and search results are instant.
6. i18n and magnific JS calls still the last things to load on page and taking up most room in Firebug timeline.
7. Confirmed via access and error logs on the server that search results are not the culprit to lag time. 
8. Results are called and populated, but the page seems to keep checking for something or constantly updating something in the background. 
9. Have removed custom styles and script files to see if results are quicker. They are not. Core WET template pieces used only (no cookies, no fancy scripts, just core WET), and still experiencing a lag. 
10. Magnific plugin still being used, but solely calling upon WET ajaxed-in content. Content is not loaded on the page until called upon.

**Comment by pjackson28** [comment]
How are you implementing the updates to the DOM? Are you inserting all the results as one single DOM insertion or you inserting each result as a separate DOM insertion? The reason I ask is the first approach would result in one reflow/repaint for all results (quickest by far) while the other would trigger a separate reflow/repaint for each result (very slow).

Also with you using addClass/removeClass, you are likely triggering a reflow/repaint on each call (very slow) so you should use one of the following approaches to avoid this:
1. Get a copy of the results in a string, add/remove classes using regex, then replace the markup in the DOM with the updated string. This would result in one DOM operation rather than many.
2. Detach the search results from the DOM, implement your add/remove, then reattach the search results to the DOM (so you only get the reflow/repaint when reattaching).
3. Manipulate the search results before inserting them into the DOM to keep to one reflow/repaint.

There may be other options, but the key is to keep your DOM manipulations to 1 or 2 at most, otherwise you are going to get huge page slowdowns when you have a lot of results.

**Comment by markbelanger** [comment]
Thanks Paul. All results are returned as a whole, once they've been returned from the request to the server. The page is built as one. The records are not built and sent separately. 

I think it might be caused by the fact that I'm using panels as a container for each result, which in itself contains 2 or 3 panels (depends on user preferences) containing the record's content in each language (1 panel per language). 

Because I'm allowing the user to switch the order and number of languages on-the-fly, I think this is where it might be getting bogged down. 

Even though the results are returned as one page, the cookies are getting and setting each panel with each record when the styles are applied on page load. 

The current production version of the application uses one single table, where the results are split by rows. 

I wanted to take advantage of the WET panels to easily split and identify the records and their content, based on linear semantics of the page. 

I might need to resort to a data table instead. The problem there, is I won't be able to offer the user the ability to switch column order or number without dipping back into potential jquery UI.

**Comment by pjackson28** [comment]
I'm not really clear on what features you are trying to offer. When you say switching column order or number, what do you mean by that? Is that a form of sorting or showing/hiding of columns or rows? Could you provide an example of the resulting markup?

**Comment by markbelanger** [comment]
Basically, the user has the choice of setting their results to be displayed in the two official languages, or add Spanish or Portuguese as the third language. They can also determine if they want English or French to be displayed as their first column.

To get an idea of how it works right now in production, take a look at the following page (Note: it`s old and still CLF2, you been warned :o)), where the user sets his preferences: http://www.btb.termiumplus.gc.ca/tpv2alpha/mny-termium-eng.html?lang=eng&srchtxt=&i=&index=alt (See section titled: Select the order of the display of languages). 

With the new WET4 version, I'm trying to achieve this on-the-fly, instead of having the user change pages to go change their preferences.

**Comment by pjackson28** [comment]
Where would the user be changing the column order or adding the extra language? Would it still be in a preferences screen? Couldn't you just refresh the results after the user has made the selection as to the column order? I'm assuming that column order changes and adding new languages is not that common, or at least usually not done more than once per session with a user.

**Comment by pjackson28** [comment]
Just noticed that you wanted on the fly. Question still applies as to why you couldn't just refresh the table results after the user change.

**Comment by markbelanger** [comment]
With the new interface, I'm allowing the user to always have access to the Preferences button (client business requirement). 

Here's a screenshot of how it appears to the user:
![tplus-preferences](https://cloud.githubusercontent.com/assets/3777175/8881340/0119969c-320d-11e5-9590-fe25473e73ab.png)

When selecting one of the options (onclick), the columns change on-the-fly as soon as the user has clicked them. This is now quick with the addClass/removeClass instead of switchClass. 

The issue I think is adding load time, is the fact that the cookies are read upon page load, and setting the order and number of columns for all the panels available in the results. This is where I think it`s slowing down. Not when the user changes their preferences. Moreso the render of the page based on the set cookies. 

When there are less results, it's pretty quick, but the more results there are, the longer it takes for the page to be done doing what it needs to do. Even if I can see that the results are all on the page (I can tell by the length of the scroll bar that all the results have loaded). Something is still calculating afterwards.

**Comment by pjackson28** [comment]
How are you ordering the columns before inserting into the DOM? Ideally you would be checking the required column preference upon load once, and either requesting the results in the correct column order (ideal) or building the correct column order in a string before inserting into the DOM (better than doing column reordering in the DOM). Also since this type of change should be relatively rare (or at least rarely repeated by the same user), then you could just re-request the results in the correct order or to correct the order outside of the DOM (e.g., in a string) and then reinsert into the DOM.

Since changing the column order and adding columns would be pretty rare (at least not repeated more than once per session), I'd suggest doing a refresh of the results rather than doing a complicated non-refresh approach. I'd also suggestIt should be much quicker to have the columns in the correct order before inserting into the DOM rather than trying to support switching on the fly..

**Comment by markbelanger** [comment]
Update: I took your advice in removing the on-the-fly feature for the Preferences. I agree that it's not something the user should have the need to do more than once, after first setting their preferences in the application. All traces of the on-the-fly feature (CSS, cookies) have been removed. The response and performance have indeed improved a fair deal. Results are returned quicker than the previous set of changes. There is however still a lag the more results there are on a page. Will continue to investigate.

**Comment by hsrudnicki** [comment]
@markbelanger @pjackson28 @LaurentGoderre : Are you closer to make a decision to close this issue or otherwise, please update the status of this issue?

**Comment by markbelanger** [comment]
Issue is still unresolved. Was on vacation for the past 2 weeks. Will continue to investigate this week. Next thought/step is to look into breaking the results into sets of 10 per page (optimal load time), adding pagination navigation. This will allow for results to not slow down or timeout the page, when more than 10 results are loaded. This is where the load slows down. To be continued.

**Comment by markbelanger** [comment]
Finally figured out the culprit of the lag! TABS!!! 

I had setup a tabbed interface, which offered the user access to be able to either filter their results by Subject Field or by Alphabetical Index -- the content of these change depending on the search executed by the user. 

The only way I was able to come to this conclusion, was by mistakenly breaking the tabbed interface script, when trying to implement the pagination solution (was also trying infinite scroll) I had thought would solve the issue. 

In turn, I've changed the tabs to buttons, which open the Subject Fields and Index in a left-panel overlay instead. The outcome is smooth, and best of all, resolves the lag issue 100%.

To be noted for other developers using the tabbed interface. For some reason, the more content you have in a page, the slower the load will be, when using tabs. It might be due to my Subject Fields and Index tabs having forms within each of them, I'm not sure. 

Before closing this issue, I'll leave it open, in case any of you want to add comments to my findings.

Thanks again for your help on this.

**Comment by pjackson28** [comment]
How many tabs were being created when you were still using tabs? If there were a huge number then the lag could potentially be from all the DOM hits when adding all the tab markup (adding WAI-ARIA attributes, etc).

Perhaps this is a sign that the tabs markup could be more efficient (e.g., make the changes in a string then push the updated HTML to the DOM only once or to detach the area being updated before making the changes). Same for any other plugin that makes DOM changes upon load.

**Comment by markbelanger** [comment]
There were only (and always) 3 tabs being loaded:
1. Search results
2. Filter by Subject Field
3. Alphabetical list of terms

Since the lag was happening when many results were returned in the 1st tab (2nd and 3rd tabs are always a maximum of 20 list items), I'm thinking the page calculation to compare to the other tabs (with limited content) of where the tabs break in content, were constantly being calculated when scrolling (which caused a halt when scrolling the approximate equivalent of the height of tabs 2 and 3). The more results there were, the more the page lagged (still calculating??), and the scrolling halt happened in a seemingly calculated manner throughout the results. However, I'm not sure the scrolling was the only lag trigger, as the page would lag right away, even before scrolling.

From what I experienced and from what you're suggesting, it seems the Tabbed Interface plug-in (and possibly others??) are continuously calculating during the time the user spends on a given page. Which in turn, could cause significant lagging if there is a large amount of content, mixed in with using plug-in that rely on calculating the given interface and screen.

**Comment by pjackson28** [comment]
I'm not sure what would be continuously calculated or how tabs would be the culprit. Tabs does the following when initializing:

1. Inject any needed markup upon load (mainly adding WAI-ARIA attributes and generating the tablist).
2. Add event handlers to the document node for handling keyboard and mouse events.
3. Add a listener for the timerpoke event which fires off once a second.

Once that is done, it just updates class names and WAI-ARIA attributes in response to events (minimal DOM changes). Any of the visual changes are done with CSS transitions so there should be no calculation of anything, unless it is done internally to the browser as part of handling the CSS.

We you able to trace any of the lagging to any particular part in the tabs JS?

**Comment by markbelanger** [comment]
I'm not quite sure either, but that was my guess, given the page halt when scrolling > which no longer happens without the tabs. 

I wan't able to see exactly where in the tabs the JS was causing the lagging, but from my previous comments in this issue (see screenshots above), the Firebug lagging was taking most of the time during the Magnific pop-up and the i18n handling. Could there be a conflict between those elements, when using tabs as well? Do they share any properties or do some of their properties conflict?

**Comment by pjackson28** [comment]
Looking at your previous results, I noticed a big bottleneck in gtm.js, which looking at the bars, could lead to blocking for the i18n file (since in several of the screenshots en.js loading starts after gtm.js and finishes shortly after gtm.js loading finishers). I have seen major sluggishness caused by Google Tag Manager in the past when it is tagging too much or is configured incorrectly. Did you still get the lag when you disabled Google Tag Manager?

Another thing to consider is the i18n file is loaded before any of the tabs code executes (the finishing of loading of the i18n file triggers a timerpoke event which is the event that triggers any plugin work). So if there is bottlenecking in the i18n file or before then there shouldn't be any involvement from tabs.

**Comment by pjackson28** [comment]
Note also that any delay in the i18n file loading will result in a direct delay to any of the plugins loading as none of the plugins execute until the i18n file loading triggers a timerpoke event (through calling wb.start()). So if gtm.js is blocking the i18n file from loading (or if something else is doing that) then everything else would be delayed as well by the amount the i18n file load was delayed.

**Comment by markbelanger** [comment]
Yes, I removed the GTM completely during testing, and the results were nearly identical in the lagging duration. 

I understand what you mean about the trigger and the i18n, but for some reason, it was hanging only when using tabs. 

The trigger is not blocking with the many other plug-ins I'm using. 

Is there a call back to the i18n for every tab, or is the tabbed interface built after the i18n is done and the trigger is released?

**Comment by pjackson28** [comment]
The way it works is the plugin selectors used on the page are put in a central registry (for later use). Next the i18n file is downloaded and processed once (stored once in wb.i18nDict which is attached to the window object). Once that is done, the wb.start() is fired which then triggers a timerpoke event on each of the selectors in the central registry. Once the tabs plugin receives the timerpoke event, it calls wb.i18n to pull out any needed strings from wb.i18nDict and stores in variables scoped to the plugin (does this only once for all tabs instances on the page).

So the i18n file should only be downloaded once, processed once and the tabs should be pulling out the needed strings from the cached object only once (caching for the tabs plugin). So the tabs plugin makes exactly 9 calls to wb.i18n for the whole page so I don't see how i18n or tabs would be the culprits considering all the caching done to avoid multiple repeat calls.

I'm still puzzled by the long bar for gtm.js and en.js as en.js is just an object declaration and there is nothing there really to process.

**Comment by markbelanger** [comment]
Have other users experienced similar issues because of gtm.js? Based on your explanation, there should be no lagging at all, no matter which plug-ins are being used. But even when removing gtm.js, I was still getting the lag when using tabs. Not sure where the actual problem lays, since it doesn't seem to make sense based on calls, loads and triggers.

**Comment by pjackson28** [comment]
Yes, there were slowdowns due to over-aggressive tagging by gtm.js (because of what was configured). The performance got much better when the tagging was optimized.

As for why you still got the issue when removing gtm.js, I'm not sure as the WET plugin/framework itself shouldn't be the culprit due to all the internal caching. Perhaps gtm.js wasn't actually doing the blocking but instead was being blocked by something else, and when you removed gtm.js then the next script in line was getting blocked. Normally blocking is only caused by files in the head tag and wouldn't be caused by any files loaded asynchronously (such as the i18n file and JS files at the end of the body tag). Could there be anything else in the head tag that could be causing blocking in other files?

**Comment by markbelanger** [comment]
Apart from the standard wet-boew theme template package head file (line-per-line, out of the box), the one and only thing I added was the font-awesome CSS file, to support some of the icons I'm using. I removed them during my sanity checks when the lagging was going on, just in case, along with the gtm.js.

**Comment by markbelanger** [comment]
Ok, this topic can officially now be closed. 

The exact issue: 
I was taking advantage of the ajaxed-in content being called within the overlay modal plug-in. The lag was happening on pages with more results only. This is because if I had 50-100 results, this consisted in having potentially 500-1000 available sources within those results which could be called within those ajaxed-in overlays. 

Basically the page was being calculated based on how many overlays it potentially needed, and was doing the timerpoke half second delay on the page, when it wasn't even completed calculating how many overlays the page even needed. 

I tested changing the timerpoke delay, just to see if it would help, but the amount of overlays required was the key issue.

I've reverted the way the sources are being called to a simple hyperlink to a new page per available source within each record. 

I had hoped the ajaxed-in overlay content would be possible, as it allowed the user to keep his search results page under the overlay. For performance issues, this is not an option.
